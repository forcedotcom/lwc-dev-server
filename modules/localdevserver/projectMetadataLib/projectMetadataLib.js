async function loadMetadata() {
    if (window.LocalDev) {
        return Promise.resolve(window.LocalDev.project);
    } else {
        return Promise.reject(new Error('project metadata not set'));
    }
}

export async function getProjectMetadata() {
    return loadMetadata();
}

export async function getComponentMetadata(jsName, packageName) {
    return new Promise(async (resolve, reject) => {
        const metadata = await loadMetadata();
        let pkg;
        if (packageName) {
            pkg = metadata.packages.find(p => p.packageName === packageName);
        }
        if (!pkg) {
            pkg = metadata.packages.find(p => !!p.isDefault);
        }
        if (pkg) {
            const component = pkg.components.find(cmp => cmp.jsName === jsName);
            if (component) {
                resolve(component);
                return;
            } else {
                reject(`Unable to find component '${jsName}'`);
            }
            resolve(pkg);
        } else {
            reject(`Unable to find package ${packageName || 'default'}`);
        }
    });
}
