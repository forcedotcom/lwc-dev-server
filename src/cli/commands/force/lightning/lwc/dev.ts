import { flags, SfdxCommand } from '@salesforce/command';
import { Messages, SfdxError } from '@salesforce/core';
import { AnyJson, JsonArray, JsonMap } from '@salesforce/ts-types';
import * as http from 'http';
import Project from '../../../../../common/Project';
import SfdxConfiguration from '../../../../../user/SfdxConfiguration';
import LocalDevServer from '../../../../../server/LocalDevServer';
import debugLogger from 'debug';

const debug = debugLogger('localdevserver');

// Initialize Messages with the current plugin directory
Messages.importMessagesDirectory(__dirname);

// Load the specific messages for this file. Messages from @salesforce/command, @salesforce/core,
// or any library that is using the messages framework can also be loaded this way.
const messages = Messages.loadMessages('lwc-dev-server', 'dev');

export default class Dev extends SfdxCommand {
    public static description = messages.getMessage('commandDescription');

    public static examples = [
        `$ sfdx force:lighting:lwc:dev --open myComponent`
    ];

    public static args = [{ open: 'file', name: 'lwc-dev' }];

    protected static flagsConfig = {
        open: flags.string({
            char: 'o',
            description: messages.getMessage('openFlagDescription')
        }),
        port: flags.integer({
            char: 'p',
            description: messages.getMessage('portFlagDescription')
        })
    };

    // Comment this out if your command does not require an org username
    protected static requiresUsername = true;

    // Comment this out if your command does not support a hub org username
    protected static supportsDevhubUsername = true;

    // Set this to true if your command requires a project workspace; 'requiresProject' is false by default
    protected static requiresProject = true;

    public async run(): Promise<AnyJson> {
        const componentName = this.flags.open;
        const port = this.flags.port || 3333;

        if (!this.org) {
            this.ux.log('org was undefined, an org is required.');
            return { org: typeof this.org };
        }

        if (!this.project) {
            this.ux.log(
                'project is undefined. this must be run from a sfdx worksapce'
            );
            return { project: typeof this.project };
        }

        // this.org is guaranteed because requiresUsername=true, as opposed to supportsUsername
        const conn = this.org.getConnection();

        // Highest level API is always last
        const api_version = await conn.retrieveMaxApiVersion();

        debug(
            `You appear to be running on a Salesforce instance that can support up to API level ${api_version}`
        );

        if (componentName) {
            debug(`You wanted to open this component: ${componentName}`);
        }

        // TODO resolve location of `componentName`, ensure directory structure is imported / compiled

        // TODO check if it's already running on the port first

        // custom onProxyReq function to inject into Talon's proxy
        // this will insert the Authorization header to have the requests be authenticated
        const onProxyReq = function(
            proxyReq: http.ClientRequest,
            req: http.IncomingMessage,
            res: http.ServerResponse
        ) {
            proxyReq.setHeader('Authorization', `Bearer ${conn.accessToken}`);
            // req.headers.Cookie = `sid=${sid_cookie}`;
        };

        const project = new Project(this.project.getPath());

        const sfdxConfiguration = new SfdxConfiguration(project);
        sfdxConfiguration.api_version = api_version;
        sfdxConfiguration.endpoint = conn.instanceUrl;
        sfdxConfiguration.onProxyReq = onProxyReq;
        sfdxConfiguration.port = port;
        sfdxConfiguration.namespace = <string>(
            (await this.project.resolveProjectConfig()).namespace
        );

        project.sfdxConfiguration = sfdxConfiguration;

        const retValue = {
            orgId: this.org.getOrgId(),
            api_version: sfdxConfiguration.api_version,
            endpoint: sfdxConfiguration.endpoint,
            onProxyReq: JSON.stringify(sfdxConfiguration.onProxyReq),
            componentName,
            port,
            token: conn.accessToken
        };
        debug(JSON.stringify(retValue));

        // Start local dev server
        // TODO pass in component to open & open browser
        new LocalDevServer().start(project);

        return retValue;
    }
}
