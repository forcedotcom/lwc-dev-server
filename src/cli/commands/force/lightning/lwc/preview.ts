#!/usr/bin/env ts-node
import { flags, SfdxCommand } from '@salesforce/command';
import { Messages, SfdxError } from '@salesforce/core';
import { CommandLineUtils } from '../../../../../common/Common';
import { Launcher } from '../../../../../common/Requirements';
import { LaunchIOS } from '../../../../../common/LaunchIOS';
import { LaunchAndroid } from '../../../../../common/LaunchAndroid';
import { DefaultLWCMobileConfig } from '../../../../../common/Config';

// Initialize Messages with the current plugin directory
Messages.importMessagesDirectory(__dirname);

// Load the specific messages for this file. Messages from @salesforce/command, @salesforce/core,
// or any library that is using the messages framework can also be loaded this way.
const messages = Messages.loadMessages('@salesforce/lwc-dev-server', 'preview');

export default class Preview extends SfdxCommand {
    public static description = messages.getMessage('commandDescription');

    public static examples = [
        `$ sfdx force:lightning:lwc:preview -p iOS -t LWCSim2 -f http://localhost:3333`,
        `$ sfdx force:lightning:lwc:preview -p Android -t LWCEmu2 -f http://localhost:3333`
    ];

    public static args = [{ name: 'file' }];

    protected static flagsConfig = {
        // flag with a value (-n, --name=VALUE)
        platform: flags.string({
            char: 'p',
            description: messages.getMessage('platformFlagDescription')
        }),
        path: flags.string({
            char: 'f',
            description: messages.getMessage('pathFlagDescription')
        }),
        target: flags.string({
            char: 't',
            description: messages.getMessage('targetFlagDescription')
        })
    };

    // Comment this out if your command does not require an org username
    protected static requiresUsername = false;

    // Comment this out if your command does not support a hub org username
    protected static supportsDevhubUsername = false;

    // Set this to true if your command requires a project workspace; 'requiresProject' is false by default
    protected static requiresProject = false;

    public async run(): Promise<any> {
        let launcher: Launcher;
        let url: string = '';

        if (this.flags.platform === undefined) {
            return new Promise<any>((resolve, reject) => reject(false));
        }
        //FIXME: Invoke setup and use detected settings!
        if (CommandLineUtils.platformFlagIsIOS(this.flags.platform)) {
            launcher = new LaunchIOS(this.flags.target, 'iPhone-X', 'iOS-13-3');
            url = this.flags.path;
        } else if (
            CommandLineUtils.platformFlagIsAndroid(this.flags.platform)
        ) {
            launcher = new LaunchAndroid(
                this.flags.target,
                'pixel',
                'android-28'
            );
            url = this.flags.path;
        }

        // if (this.flags.platform.toLowerCase() == "ios") {
        //   launcher = new LaunchIOS(this.flags.target, "iPhone-X","iOS-13-3");
        //   url = this.flags.path;
        // } else if (this.flags.platform.toLowerCase() == "android") {
        //   launcher = new LaunchAndroid(this.flags.target,"pixel","android-28");
        //   url = this.flags.path;
        // }
        // return launcher.launchNativeBrowser(url);
        return new Promise<any>((resolve, reject) => {
            if (launcher == null) {
                reject('Could not launch target');
                return;
            }
            resolve(launcher.launchNativeBrowser(url));
        });
    }
}
