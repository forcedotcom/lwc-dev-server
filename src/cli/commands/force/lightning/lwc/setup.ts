#!/usr/bin/env ts-node

import { flags, SfdxCommand } from '@salesforce/command';
import { Messages, SfdxError } from '@salesforce/core';
import { CommandLineUtils } from '../../../../../common/Common';
import {
    RequirementList,
    SetupTestResult
} from '../../../../../common/Requirements';
import { AndroidEnvironmentSetup } from '../../../../../common/AndroidEnvironmentSetup';
import { IOSEnvironmentSetup } from '../../../../../common/IOSEnvironmentSetup';

// Initialize Messages with the current plugin directory
Messages.importMessagesDirectory(__dirname);

// Load the specific messages for this file. Messages from @salesforce/command, @salesforce/core,
// or any library that is using the messages framework can also be loaded this way.
const messages = Messages.loadMessages('@salesforce/lwc-dev-server', 'preview');

export default class Setup extends SfdxCommand {
    public static description = messages.getMessage('commandDescription');

    public static examples = [
        `$ sfdx force:lightning:lwc:setup -p iOS`,
        `$ sfdx force:lightning:lwc:setup -p Android`
    ];

    protected static flagsConfig = {
        // flag with a value (-n, --name=VALUE)
        platform: flags.string({
            char: 'p',
            description: messages.getMessage('platformFlagDescription')
        })
    };

    public async run(): Promise<SetupTestResult> {
        let setup: RequirementList;
        //let setupResult: SetupTestResult = new  SetupTestResult();

        if (this.flags.platform === undefined) {
            return new Promise((resolve, reject) => {
                reject({ hasMetAllRequirements: false, tests: [] });
            });
        }

        if (CommandLineUtils.platformFlagIsIOS(this.flags.platform)) {
            setup = new IOSEnvironmentSetup();
        } else {
            setup = new AndroidEnvironmentSetup();
        }
        return setup.executeSetup();
    }
}
