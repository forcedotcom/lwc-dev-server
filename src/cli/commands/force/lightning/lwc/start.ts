import { flags, SfdxCommand } from '@salesforce/command';
import { Messages } from '@salesforce/core';
import { AnyJson } from '@salesforce/ts-types';
import * as http from 'http';
import Project from '../../../../../common/Project';
import LocalDevServer from '../../../../../server/LocalDevServer';
import debugLogger from 'debug';
import colors from 'colors';

const debug = debugLogger('localdevserver');

// Initialize Messages with the current plugin directory
Messages.importMessagesDirectory(__dirname);

// Load the specific messages for this file. Messages from @salesforce/command, @salesforce/core,
// or any library that is using the messages framework can also be loaded this way.
const messages = Messages.loadMessages('@salesforce/lwc-dev-server', 'start');

export default class Start extends SfdxCommand {
    public static description = messages.getMessage('commandDescription');

    public static examples = [
        `$ sfdx force:lighting:lwc:start`,
        `$ sfdx force:lighting:lwc:start --port 3000`
    ];

    public static args = [];

    protected static flagsConfig = {
        port: flags.integer({
            char: 'p',
            description: messages.getMessage('portFlagDescription')
        })
    };

    // Comment this out if your command does not support specifying an org username
    protected static supportsUsername = true;

    // Comment this out if your command does not support a dev hub org username
    protected static supportsDevhubUsername = true;

    // Set this to true if your command requires a project workspace; 'requiresProject' is false by default
    protected static requiresProject = true;

    public async run(): Promise<AnyJson> {
        const defaultdevhubusername = this.configAggregator.getPropertyValue(
            'defaultdevhubusername'
        ) as string;

        let port: number;
        if (this.flags.port !== undefined && this.flags.port !== null) {
            port = this.flags.port;
        } else {
            port = 3333;
        }

        if (!this.org) {
            // This you DO need.
            // We require this right now for proxying and api version.
            // If we do not have an org, we can still function.
            //  - We should disable the proxying and then allow specification of the version
            //    if you don't authenticate.
            //

            const targetusername = this.flags.targetusername;
            if (targetusername) {
                this.reportStatus(
                    colors.green(defaultdevhubusername),
                    colors.red(
                        `${targetusername} - ${messages.getMessage(
                            'error:invalidscratchorgusername'
                        )}`
                    )
                );
            } else {
                const configuredusername = this.configAggregator.getPropertyValue(
                    'defaultusername'
                );
                this.reportStatus(
                    colors.green(defaultdevhubusername),
                    colors.red(
                        `${configuredusername} - ${messages.getMessage(
                            'error:noscratchorg'
                        )}`
                    )
                );
            }

            return { org: typeof this.org };
        }

        // Sfdx validates this before we have a chance to, this appears to be
        // a "just in case" condition so reporting the same error they do.
        if (!this.project) {
            this.ux.error(messages.getMessage('error:noproject'));
            return { project: typeof this.project };
        }

        this.ux.log(colors.gray(messages.getMessage('legal:cliusage')));

        // this.org is guaranteed because requiresUsername=true, as opposed to supportsUsername
        const conn = this.org.getConnection();

        // Highest level API is always last
        const api_version = await conn.retrieveMaxApiVersion();

        const orgusername = this.org.getUsername() || '';
        try {
            // currently something in sfdx is resulting in an unhandled
            // promise rejection, instead of surfacing that rejection
            // through the returned promise. @W-6723813
            process.on('unhandledRejection', reason => {
                if (
                    reason &&
                    // @ts-ignore
                    reason.name &&
                    // @ts-ignore
                    reason.name === 'StatusCodeError'
                ) {
                    // ignore unhandled rejects during below refresh call
                    debug(`unhandledPromiseRejection: ${reason}`);
                } else {
                    this.ux.error(`unhandledPromiseRejection: ${reason}`);
                }
            });
            await this.org.refreshAuth();
        } catch (err) {
            this.reportStatus(
                colors.green(defaultdevhubusername),
                colors.red(
                    `${orgusername} - ${messages.getMessage(
                        'error:inactivescratchorg'
                    )}`
                ),
                colors.green(api_version)
            );
            return { error: err };
        }

        this.reportStatus(
            colors.green(defaultdevhubusername),
            colors.green(orgusername),
            colors.green(api_version)
        );

        const accessToken = conn.accessToken;

        // custom onProxyReq function to inject into Talon's proxy
        // this will insert the Authorization header to have the requests be authenticated
        const onProxyReq = function(
            proxyReq: http.ClientRequest,
            req: http.IncomingMessage,
            res: http.ServerResponse
        ) {
            proxyReq.setHeader('Authorization', `Bearer ${accessToken}`);
        };

        const project = new Project(this.project.getPath());

        project.configuration.api_version = api_version;
        project.configuration.endpoint = conn.instanceUrl;
        project.configuration.onProxyReq = onProxyReq;
        project.configuration.port = port;
        project.configuration.namespace = <string>(
            (await this.project.resolveProjectConfig()).namespace
        );

        const retValue = {
            orgId: this.org.getOrgId(),
            api_version: project.configuration.api_version,
            endpoint: project.configuration.endpoint,
            onProxyReq: JSON.stringify(project.configuration.onProxyReq),
            port,
            token: accessToken
        };
        debug(JSON.stringify({ ...retValue, token: undefined }));

        // Start local dev server
        new LocalDevServer().start(project, conn);

        return retValue;
    }

    private reportStatus(
        devHubOrg: string,
        scratchOrg: string,
        apiVersion?: string
    ) {
        if (apiVersion) {
            this.ux.log(`\
Starting LWC Local Development.
    Dev Hub Org: ${devHubOrg}
    Scratch Org: ${scratchOrg}
    Api Version: ${apiVersion}\
`);
        } else {
            this.ux.log(`\
Starting LWC Local Development.
    Dev Hub Org: ${devHubOrg}
    Scratch Org: ${scratchOrg}\
`);
        }
    }
}
