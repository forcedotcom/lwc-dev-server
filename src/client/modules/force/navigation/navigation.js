/**
 * Provide an empty stub for force/navigation for components that use lightning/navigation.
 * Modified from webruntime_navigation/lightningNavigation
 */
import { register } from 'wire-service';

export const CurrentPageReference = Symbol('CurrentPageReference');
register(CurrentPageReference, target => {});

const Navigate = Symbol('Navigate');
const GenerateUrl = Symbol('GenerateUrl');

export const NavigationMixin = Base => {
    if (typeof Base.prototype.dispatchEvent !== 'function') {
        throw new TypeError(`${Base} must be an Element type`);
    }
    return class extends Base {
        [Navigate](pageReference, replace) {}
        [GenerateUrl](pageReference) {
            return Promise.resolve('');
        }
    };
};
NavigationMixin.Navigate = Navigate;
NavigationMixin.GenerateUrl = GenerateUrl;
