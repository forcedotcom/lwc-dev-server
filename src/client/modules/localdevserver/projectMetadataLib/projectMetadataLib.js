async function loadMetadata() {
    if (window.LocalDev) {
        return Promise.resolve(window.LocalDev.project);
    } else {
        return Promise.reject(new Error('project metadata not set'));
    }
}

export function getNonce() {
    const meta = document.head.querySelector('meta[name=nonce][content]');

    return meta ? meta.content : '';
}

export async function getProjectMetadata() {
    return loadMetadata();
}

export async function getComponentMetadata(jsName, packageKey) {
    return new Promise(async (resolve, reject) => {
        const metadata = await loadMetadata();
        let pkg;
        if (packageKey) {
            pkg = metadata.packages.find(p => p.key === packageKey);
        } else {
            pkg = metadata.packages.find(p => !!p.isDefault);
        }

        if (pkg) {
            const component = pkg.components.find(cmp => cmp.jsName === jsName);
            if (component) {
                resolve(component);
                return;
            } else {
                reject(new Error(`Unable to find component '${jsName}'`));
            }
            resolve(pkg);
        } else {
            reject(
                new Error(`Unable to find package ${packageName || 'default'}`)
            );
        }
    });
}
