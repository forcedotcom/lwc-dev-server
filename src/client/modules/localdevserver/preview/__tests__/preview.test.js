import Preview from 'localdevserver/preview';
import { createElement } from 'lwc';
import { getComponentMetadata } from 'localdevserver/projectMetadataLib';
import { subscribe } from 'webruntime_navigation/navigation';
import {
    flushPromises,
    mockComponentMetadata
} from '../../../__tests__/testutils';

jest.mock('localdevserver/projectMetadataLib');

jest.mock('webruntime_navigation/navigation', () => ({
    navigate: jest.fn(),
    generateUrl: jest.fn(() => Promise.resolve('/mock/url')),
    subscribe: jest.fn((context, callback) => {
        callback();
        return { unsubscribe: jest.fn() };
    }),
    getNavigationContextId: jest.fn()
}));

jest.mock('wire-service', () => ({
    register: jest.fn((symbol, wireCallback) => {
        wireCallback({
            addEventListener(eventName, callback) {
                callback();
            },
            dispatchEvent: jest.fn()
        });
    }),
    LinkContextEvent: jest.fn(),
    ValueChangedEvent: jest.fn()
}));

function createComponentUnderTest(props) {
    const el = createElement('localdevserver-preview', { is: Preview });
    Object.assign(el, props);
    document.body.appendChild(el);
    return el;
}

describe('preview', () => {
    let subscriptionMock;
    let consoleLogMock;
    let consoleWarnMock;
    let consoleErrorMock;
    let consoleGroupMock;

    beforeEach(() => {
        subscriptionMock = {
            unsubscribe: jest.fn()
        };

        consoleLogMock = jest.spyOn(console, 'log').mockImplementation();
        consoleWarnMock = jest.spyOn(console, 'warn').mockImplementation();
        consoleErrorMock = jest.spyOn(console, 'error').mockImplementation();
        consoleGroupMock = jest.spyOn(console, 'group').mockImplementation();
    });

    afterEach(() => {
        jest.resetAllMocks();
        consoleLogMock.mockRestore();
        consoleWarnMock.mockRestore();
        consoleErrorMock.mockRestore();
        consoleGroupMock.mockRestore();
    });

    it('renders the component', async () => {
        getComponentMetadata.mockImplementation(() => {
            return Promise.resolve(mockComponentMetadata('c', 'button'));
        });

        const route = {
            attributes: {
                namespace: 'c',
                name: 'button'
            }
        };

        subscribe.mockImplementation((context, callback) => {
            callback(route);
            return subscriptionMock;
        });

        const element = createComponentUnderTest();

        await flushPromises();
        expect(element).toMatchSnapshot();
    });

    it('should show a loading indicator', async () => {
        getComponentMetadata.mockImplementation(() => {
            return Promise.resolve(mockComponentMetadata('c', 'button'));
        });

        const route = {
            attributes: {
                namespace: 'c',
                name: 'button'
            }
        };

        subscribe.mockImplementation((context, callback) => {
            callback(route);
            return subscriptionMock;
        });

        const element = createComponentUnderTest();

        let spinner = element.shadowRoot.querySelector('lightning-spinner');
        expect(spinner).toBeTruthy();

        await flushPromises();

        spinner = element.shadowRoot.querySelector('lightning-spinner');
        expect(spinner).toBeFalsy();
    });

    it('should show an error message if the route is missing the attributes', async () => {
        getComponentMetadata.mockImplementation(() => {
            return Promise.resolve(mockComponentMetadata('c', 'button'));
        });

        const route = {
            state: {
                namespace: 'c',
                name: 'button'
            }
        };

        subscribe.mockImplementation((context, callback) => {
            callback(route);
            return subscriptionMock;
        });

        const element = createComponentUnderTest();
        await flushPromises();

        const errorEl = element.shadowRoot.querySelector(
            'localdevserver-error'
        );
        expect(errorEl).toBeTruthy();

        const errorText = errorEl.shadowRoot.textContent;
        expect(errorText).toContain(
            'The component to preview was not specified'
        );
    });

    it('should show an error message if the route has invalid attributes', async () => {
        getComponentMetadata.mockImplementation(() => {
            return Promise.resolve(mockComponentMetadata('c', 'button'));
        });

        const route = {
            attributes: {
                specifier: 'c/button'
            }
        };

        subscribe.mockImplementation((context, callback) => {
            callback(route);
            return subscriptionMock;
        });

        const element = createComponentUnderTest();
        await flushPromises();

        const errorEl = element.shadowRoot.querySelector(
            'localdevserver-error'
        );
        expect(errorEl).toBeTruthy();

        const errorText = errorEl.shadowRoot.textContent;
        expect(errorText).toContain(
            'The component to preview was not specified'
        );
    });

    it('should show an error message if the component is not found in the metadata', async () => {
        getComponentMetadata.mockImplementation(() => {
            throw new Error('not found');
        });

        const route = {
            attributes: {
                namespace: 'c',
                name: 'foo'
            }
        };

        subscribe.mockImplementation((context, callback) => {
            callback(route);
            return subscriptionMock;
        });

        const element = createComponentUnderTest();
        await flushPromises();

        const errorEl = element.shadowRoot.querySelector(
            'localdevserver-error'
        );
        expect(errorEl).toBeTruthy();

        const errorText = errorEl.shadowRoot.textContent;
        expect(errorText).toContain(
            `The component named 'c/foo' was not found`
        );
    });

    it.todo('should show an error message if the component has compile errors');

    it('should call unsubscribe when disconnected', () => {
        getComponentMetadata.mockImplementation(() => {
            return Promise.resolve(mockComponentMetadata('c', 'button'));
        });

        const route = {
            attributes: {
                namespace: 'c',
                name: 'button'
            }
        };

        subscribe.mockImplementation((context, callback) => {
            callback(route);
            return subscriptionMock;
        });

        const element = createComponentUnderTest();
        element.parentNode.removeChild(element);

        expect(subscriptionMock.unsubscribe).toBeCalledTimes(1);
    });
});
