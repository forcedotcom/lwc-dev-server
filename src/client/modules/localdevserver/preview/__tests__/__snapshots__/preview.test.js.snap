// Jest Snapshot v1, https://goo.gl/fbAQLP

// exports[`preview displays errors 1`] = `
// <localdevserver-preview>
//   #shadow-root(open)
//     <localdevserver-layout-section
//       class="main-content"
//     >
//       #shadow-root(open)
//         <slot />
//       <article
//         class="slds-card"
//       >
//         <div
//           class="header slds-grid slds-grid_vertical-align-center slds-grid_align-spread slds-p-vertical_medium slds-p-horizontal_large"
//         >
//           <header>
//             <lightning-breadcrumbs>
//               #shadow-root(open)
//                 <slot />
//               <lightning-breadcrumb>
//                 #shadow-root(open)
//               </lightning-breadcrumb>
//               <lightning-breadcrumb>
//                 #shadow-root(open)
//               </lightning-breadcrumb>
//             </lightning-breadcrumbs>
//             <h2
//               class="component-name slds-card__header-title slds-truncate slds-text-heading_medium slds-m-top_small"
//               title="c-foo"
//             >
//               c-foo
//             </h2>
//           </header>
//           <div
//             class="header-buttons"
//           >
//             <a
//               class="slds-button slds-button_outline-brand"
//               href="/"
//             >
//               Go Back
//             </a>
//             <a
//               class="slds-button slds-button_brand"
//               href="vscode://file//Users/arya/dev/test/src/foo/foo.js"
//             >
//               View in VS Code
//             </a>
//           </div>
//         </div>
//         <div
//           class="preview-content slds-grid slds-grid_vertical-align-center slds-grid_align-center"
//         >
//           <div
//             class="container slds-scrollable slds-m-around_large"
//           />
//           <localdevserver-error>
//             #shadow-root(open)
//               <section
//                 aria-modal="true"
//                 class="slds-modal slds-fade-in-open slds-modal_large"
//                 role="dialog"
//                 tabindex="-1"
//               >
//                 <div
//                   class="slds-modal__container"
//                 >
//                   <header
//                     class="slds-modal__header"
//                   >
//                     <h2
//                       id="modal-heading-01-19"
//                     >
//                       <img
//                         class="error-image"
//                         src="/assets/images/buggy.svg"
//                       />
//                     </h2>
//                   </header>
//                   <div
//                     class="slds-modal__content slds-p-around_medium"
//                   >
//                     <h1
//                       class="error-message slds-text-heading_large"
//                     />
//                     <h2
//                       class="error-location slds-text-heading_small"
//                     />
//                     <localdevserver-code-highlighter
//                       class="error-code"
//                     >
//                       #shadow-root(open)
//                         <div />
//                     </localdevserver-code-highlighter>
//                     <localdevserver-error-stacks
//                       class="error-stacks"
//                     >
//                       #shadow-root(open)
//                         <div
//                           class="stack-message slds-text-heading_small"
//                         >
//                           <div
//                             class="button-container"
//                           >
//                             <div
//                               class="line"
//                             />
//                             <div
//                               class="collapse-container"
//                             >
//                               <div
//                                 class="collapse"
//                               >
//                                 0 stack frames collapsed
//                               </div>
//                             </div>
//                           </div>
//                         </div>
//                     </localdevserver-error-stacks>
//                     <div
//                       class="info-note slds-text-heading_small"
//                     >
//                       This screen is visible only in development. It will not appear if the app crashes in production. Open your browser's developer console to further inspect this error.
//                     </div>
//                   </div>
//                 </div>
//               </section>
//               <div
//                 class="slds-backdrop slds-backdrop_open"
//               />
//           </localdevserver-error>
//         </div>
//       </article>
//     </localdevserver-layout-section>
// </localdevserver-preview>
// `;

exports[`preview renders 1`] = `
<localdevserver-preview>
  #shadow-root(open)
    <localdevserver-layout-section
      class="main-content"
    >
      #shadow-root(open)
        <slot />
      <article
        class="slds-card"
      >
        <div
          class="header slds-grid slds-grid_vertical-align-center slds-grid_align-spread slds-p-vertical_medium slds-p-horizontal_large"
        >
          <header>
            <lightning-breadcrumbs>
              #shadow-root(open)
                <slot />
              <lightning-breadcrumb>
                #shadow-root(open)
              </lightning-breadcrumb>
              <lightning-breadcrumb>
                #shadow-root(open)
              </lightning-breadcrumb>
            </lightning-breadcrumbs>
            <h2
              class="component-name slds-card__header-title slds-truncate slds-text-heading_medium slds-m-top_small"
            />
          </header>
          <div
            class="header-buttons"
          >
            <a
              class="slds-button slds-button_outline-brand"
              href="/"
            >
              Go Back
            </a>
            <a
              class="slds-button slds-button_brand"
              href="javascript:void(0);"
            >
              View in VS Code
            </a>
          </div>
        </div>
        <div
          class="preview-content slds-grid slds-grid_vertical-align-center slds-grid_align-center"
        >
          <div
            class="container slds-scrollable slds-m-around_large"
          />
          <lightning-spinner>
            #shadow-root(open)
          </lightning-spinner>
        </div>
      </article>
    </localdevserver-layout-section>
</localdevserver-preview>
`;

// exports[`preview renders a component 1`] = `
// <localdevserver-preview>
//   #shadow-root(open)
//     <localdevserver-layout-section
//       class="main-content"
//     >
//       #shadow-root(open)
//         <slot />
//       <article
//         class="slds-card"
//       >
//         <div
//           class="header slds-grid slds-grid_vertical-align-center slds-grid_align-spread slds-p-vertical_medium slds-p-horizontal_large"
//         >
//           <header>
//             <lightning-breadcrumbs>
//               #shadow-root(open)
//                 <slot />
//               <lightning-breadcrumb>
//                 #shadow-root(open)
//               </lightning-breadcrumb>
//               <lightning-breadcrumb>
//                 #shadow-root(open)
//               </lightning-breadcrumb>
//             </lightning-breadcrumbs>
//             <h2
//               class="component-name slds-card__header-title slds-truncate slds-text-heading_medium slds-m-top_small"
//               title="c-foo"
//             >
//               c-foo
//             </h2>
//           </header>
//           <div
//             class="header-buttons"
//           >
//             <a
//               class="slds-button slds-button_outline-brand"
//               href="/"
//             >
//               Go Back
//             </a>
//             <a
//               class="slds-button slds-button_brand"
//               href="vscode://file//Users/arya/dev/test/src/foo/foo.js"
//             >
//               View in VS Code
//             </a>
//           </div>
//         </div>
//         <div
//           class="preview-content slds-grid slds-grid_vertical-align-center slds-grid_align-center"
//         >
//           <div
//             class="container slds-scrollable slds-m-around_large"
//           >
//             <div>
//               test element
//             </div>
//           </div>
//         </div>
//       </article>
//     </localdevserver-layout-section>
// </localdevserver-preview>
// `;
